Disable diff output when using -q flag in uncrustify.

smavridis opened the pull request:
Hi,  

I recently started using your pre-commit hooks and  specifically the uncrustify hook.
The only thing that bothered me, was the diff output as it would 'spam' the terminal.
In this pull request i have used the existing uncrustify ``quiet`` flag ``-q`` to also disable the diff output.

If you are interested i could also, look at adding same/similar flags to the other formaters.

I have tested it in a private repo, but have not (yet) added unitests as i couldn't run the pytest due to requirements.

Let me know what you think of the changes/feature, and what i could add/improve.

Thank you for your work

pocc commented:
This is a fine idea. You would need to do the following:

1. Ensure it works for all formatters
1. Add unit tests that pass on travis/circleci
1. Add documentation about this flag to the README (section after https://github.com/pocc/pre-commit-hooks#enforcing-linter-version-with---version)

I'm happy to help you with environment setup if you have questions. It looks like there are some failing formatter tests I need to take care of (likely version mismatch between macos/linux), so feel free to ignore those ones.

pocc commented on hooks/uncrustify.py:
```diff
@@ -20,6 +20,7 @@ def __init__(self, args):
         self.check_installed()
         self.parse_args(args)
         self.file_flag = "-f"
+        self.show_diff = "-q" not in self.args
```
I prefer to keep the existing convention of \*_flag. So name it `quiet_flag`, and set it to `"q"`.

pocc commented on hooks/utils.py:
```diff
@@ -132,6 +132,7 @@ class FormatterCmd(Command):
     def __init__(self, command, look_behind, args):
         super().__init__(command, look_behind, args)
         self.file_flag = None
+        self.show_diff = True
```
Change name of var and to `None` as well.

smavridis replied:
Done in 652bda6

pocc commented:
You probably want to parse the `-q`/`--quiet` arg in utils.py > Command > parse_args. This would also mean you would want to have this flag as an attribute of the Command class. Given that your pull request only modifies stderr in a specific place and not globally, it's not really "quiet" per se as much as "suppress verbose output" (maybe move the -q check to [the writing of stderr](https://github.com/pocc/pre-commit-hooks/blob/f7951acb030653d1e19279366cb72800ca470dac/hooks/uncrustify.py#L47) in each command's run function)?

smavridis commented:
> You probably want to parse the `-q`/`--quiet` arg in utils.py > Command > parse_args. This would also mean you would want to have this flag as an attribute of the Command class. Given that your pull request only modifies stderr in a specific place and not globally, it's not really "quiet" per se as much as "suppress verbose output" (maybe move the -q check to [the writing of stderr](https://github.com/pocc/pre-commit-hooks/blob/f7951acb030653d1e19279366cb72800ca470dac/hooks/uncrustify.py#L47) in each command's run function)?

I now realize there are a few ambiguities in what i/we want to do with the flags.

My initial intention was to do two things:
 - Disable the diff generated by your hooks.
 - Silence uncrustify.

I now think, that it was wrong to control both things with the same flag as they should be user selectable and independent.

So perhaps we should use a flag that does not conflict with any existing flag of any formater and only hides the hook generated diffs?

Tomorrow i will setup my dev environment, i guess Ubuntu is the preferred distribution for running the tests?

pocc commented:
The tests should honestly run well on every platform.

smavridis commented:
Hi again, real life got in the way....

I made the changes i mentioned in my last message, (bcc9582d3c6c38c3d09569aa98284156f1bce74b)
I also added unit tests(scenarios?) to stress the new flag(2ec6d269977a300f0b4cfc73b8c4a552b08c79e3).
I have seen the unit test code, and from what i understood, you create scenarios, and these scenarios get 'converted' into two kind of tests:

- shell_cmd: basically run the native executable
- cmd_class: This creates the respective class, and uses it to execute the command.

Currently both tests expect to have exactly the same behavior, as they are created from a single scenario (get_multifile_scenarios_no_diff() in my case)
However the flag i added breaks this assumption(intentionally).

Now what i see, is that the shell_cmd tests fail.
This given my understanding is to be expected, as the scenario i wrote (that uses --no-diff) assumes that no output will be printed.
This however wont work for the shell_cmd, as --no-diff is only recognized by the hooks,  and is not a 'native' flag for clag-format/uncrustify commands. 

What is your take on all this?
Thanks for your time, and patience ;-D
